firedrake/tsfc_interface.py

```diff
@@ -52,75 +47,33 @@
-        # don't get deadlocks. But MPI_Comm objects are not hashable,
-        # so use comm.py2f() since this is an internal communicator and
-        # hence the C handle is stable.
-        commkey = comm.py2f()
-        assert commkey != MPI.COMM_NULL.py2f()
-        return cls._cache.get((key, commkey)) or cls._read_from_disk(key, comm)
-
-    @classmethod
-    def _read_from_disk(cls, key, comm):
-        if comm.rank == 0:
-            cache = cls._cachedir
-            shard, disk_key = key[:2], key[2:]
-            filepath = os.path.join(cache, shard, disk_key)
-            val = None
-            if os.path.exists(filepath):
-                try:
-                    with gzip.open(filepath, 'rb') as f:
-                        val = f.read()
-                except zlib.error:
-                    pass
-
-            comm.bcast(val, root=0)
-        else:
-            val = comm.bcast(None, root=0)
+_cachedir = environ.get(
+    'FIREDRAKE_TSFC_KERNEL_CACHE_DIR',
+    path.join(tempfile.gettempdir(), f'firedrake-tsfc-kernel-cache-uid{getuid()}')
+)
 
-        if val is None:
-            raise KeyError(f"Object with key {key} not found")
-        return cls._cache.setdefault((key, comm.py2f()), pickle.loads(val))
 
-    @classmethod
-    def _cache_store(cls, key, val):
-        key, comm = key
-        cls._cache[(key, comm.py2f())] = val
-        _ensure_cachedir(comm=comm)
-        if comm.rank == 0:
-            val._key = key
-            shard, disk_key = key[:2], key[2:]
-            filepath = os.path.join(cls._cachedir, shard, disk_key)
-            tempfile = os.path.join(cls._cachedir, shard, "%s_p%d.tmp" % (disk_key, os.getpid()))
-            # No need for a barrier after this, since non root
-            # processes will never race on this file.
-            os.makedirs(os.path.join(cls._cachedir, shard), exist_ok=True)
-            with gzip.open(tempfile, 'wb') as f:
-                pickle.dump(val, f, 0)
-            os.rename(tempfile, filepath)
-        comm.barrier()
-
-    @classmethod
-    def _cache_key(cls, form, name, parameters, coefficient_numbers, constant_numbers, interface, diagonal=False):
-        return md5((form.signature() + name
-                    + str(sorted(parameters.items()))
-                    + str(coefficient_numbers)
-                    + str(constant_numbers)
-                    + str(type(interface))
-                    + str(diagonal)).encode()).hexdigest(), form.ufl_domains()[0].comm
+def tsfc_compile_form_hashkey(form, prefix, parameters, interface, diagonal, log):
+    # Drop prefix as it's only used for naming and log
+    # JBTODO: Can't drop prefix as tests/slate/test_optimise.py::test_partially_optimised fails, investigate
+    # it looks like the prefix is being used to create different subkernels, which conflicts with the docstring below
+    return default_parallel_hashkey(form.signature(), prefix, parameters, interface, diagonal)
```

Comment:
not sure I understand the change. Why are we allowed to drop `coefficient_numbers`?

Reply:
`coefficient_numbers` is actually derived directly from `form` so it doesn't need to be included. Also this cache now sits at a slightly different point in the stack so `coefficient_numbers` isn't even an argument to the wrapped function.